pkg/envfuncs/provider_funcs.go

```diff
@@ -0,0 +1,272 @@
+	"sigs.k8s.io/e2e-framework/klient"
+	"sigs.k8s.io/e2e-framework/klient/k8s/resources"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/klient/wait/conditions"
+	"sigs.k8s.io/e2e-framework/pkg/env"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/support"
+	"sigs.k8s.io/e2e-framework/support/kind"
+)
+
+type clusterNameContextKey string
+
+// Following placeholder interface types are defined purely to make the pointer receiver work seemlessly
+// for the generic handlers. Without this, making the pointer receiver is really painful or we have to
+// turn providers into value receivers and that doesn't really work well for the current implementation
+// we have for kind provider. xref: https://t.ly/GDxOB
+
+// Every time you add a new provider that implements the support.E2EClusterProvider interface, that
+// needs to be listed here in this interface type with properly or'ing that with the existing
+// values.
+//
+//	type provider interfae {
+//			kind.Cluster | k3d.Cluster | kwok.Cluster
+//	}
+type provider interface {
+	kind.Cluster
+}
+
+type providerRef[T any] interface {
+	*T
+	support.E2EClusterProvider
+}
+
+// GetClusterFromContext helps extract the E2EClusterProvider object from the context.
+// This can be used to setup and run tests of multi cluster e2e Prioviders.
+func GetClusterFromContext(ctx context.Context, clusterName string) (support.E2EClusterProvider, bool) {
+	c := ctx.Value(clusterNameContextKey(clusterName))
+	if c == nil {
+		return nil, false
+	}
+	cluster, ok := c.(support.E2EClusterProvider)
+	return cluster, ok
+}
+
+// Deprecated: This handler has been deprecated in favor of GetClusterFromContext
+func GetKindClusterFromContext(ctx context.Context, clusterName string) (*kind.Cluster, bool) {
+	provider, ok := GetClusterFromContext(ctx, clusterName)
+	if ok {
+		return provider.(*kind.Cluster), ok
+	}
+	return nil, ok
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateCluster which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindCluster(clusterName string) env.Func {
+	return CreateCluster[kind.Cluster](clusterName)
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateClusterWithConfig which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindClusterWithConfig(clusterName, image, configFilePath string) env.Func {
+	return CreateClusterWithConfig[kind.Cluster](clusterName, configFilePath, kind.WithImage(image))
+}
+
+// Deprecated: This handler has been deprecated in favor of DestroyCluster
+func DestroyKindCluster(name string) env.Func {
+	return DestroyCluster(name)
+}
+
+// Deprecated: This handler has been deprecated in favor of ExportClusterLogs
+func ExportKindClusterLogs(name, dest string) env.Func {
+	return ExportClusterLogs(name, dest)
+}
```

Comment:
I would keep these in the kind_funcs.go file

Reply:
I have kept this here for compatibility purpose alone as someone might already have this import in their workflow. Moving this would break that. Can we retain this here for now and eventually get rid of it down the line instead of moving this to kind_funcs.go entirely ?

Reply:
@harshanarayana I meant move it to its own file envfuncs/kind_funcs.go as before, in the same packge.

Reply:
Done